src/sysext/sysext.c

```diff
@@ -689,186 +691,384 @@ static int work_dir_for_hierarchy(const char *hierarchy, const char *resolved_up
         return 0;
 }
 
-static int use_hierarchy_as_lowest_layer(
-                const char *hierarchy,
-                const char *maybe_resolved_upper_dir,
-                char** ret_hierarchy_as_lowest_layer) {
-        _cleanup_free_ char *resolved_hierarchy = NULL;
-        int r;
+struct overlayfs_paths {
+        char *hierarchy;
+        char *resolved_hierarchy;
+        char *resolved_upper_dir;
+
+        /* NULL if merged fs is read-only */
+        char *ofs_upper_dir;
+        /* NULL if merged fs is read-only */
+        char *ofs_work_dir;
+        /* lowest index is top lowerdir, highest index is bottom lowerdir */
+        char **ofs_lower_dirs;
+        size_t ofs_lower_dirs_size;
```

Comment:
we usually use NULL terminate arrays of strings, and call that an "strv". There's a rich set of helpers to deal with this in strv.h. Hence why keep the number of entries here?

Reply:
To cheaply reduce complexity of appending to the strv. It's not used for anything else.

Reply:
this is not performance critical. I'd always focus on simplicity and easiness to read rather than optimization. Hence I'd just drop the size field here.

Reply:
Dropped the size field and both `consume_layer` and `append_layer` functions in favor of `strv_consume` and `strv_extend`.